src/OpenTelemetry/ConcurrencyModes.cs

```diff
@@ -0,0 +1,36 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+namespace OpenTelemetry;
+
+/// <summary>
+/// Describes the concurrency mode of an OpenTelemetry extension.
+/// </summary>
+[Flags]
+public enum ConcurrencyModes : byte
+{
+    /*
+    0 0 0 0 0 0 0 0
+    | | | | | | | |
+    | | | | | | | +--- Reentrant
+    | | | | | | +----- Multithreaded
+    | | | | | +------- (reserved)
+    | | | | +--------- (reserved)
+    | | | +----------- (reserved)
+    | | +------------- (reserved)
+    | +--------------- (reserved)
+    +----------------- (reserved)
+    */
+
+    /// <summary>
+    /// Reentrant, the component can be invoked recursively without resulting
+    /// a deadlock or infinite loop.
+    /// </summary>
+    Reentrant = 0b1,
```

Comment:
We removed `Global` for now should we also remove `Reentrant`? Doesn't seem to be used at the moment.

Reply:
My gut feeling is no. Multithreaded and Reentrant are closely related, the existing implementation (which uses `lock(obj)`) implies that the exporter supports reentrancy but not multithreading (which is buggy IMHO).

Reply:
No strong feelings though, I can scope this out and add it later once we figured out how to correctly handle reentrancy.

Reply:
Interesting. So if the attribute is not specified and defaults to 0 our implied default is essentially `SingleThreaded | Reentrant`?

Should the defined `Reentrant` flag then negate the default behavior?

```csharp
public enum ConcurrencyModes : byte
{
    /// <summary>
    /// Nonreentrant, the component cannot be invoked recursively without resulting
    /// in a deadlock or infinite loop.
    /// </summary>
    Nonreentrant = 0b1,

    /// <summary>
    /// Multithreaded, the component can be invoked concurrently across
    /// multiple threads.
    /// </summary>
    Multithreaded = 0b10,
}
```

Reply:
> Interesting. So if the attribute is not specified and defaults to 0 our implied default is essentially `SingleThreaded | Reentrant`?

Maybe. I personally consider this as a bug in the current SDK implementation.

> Should the defined `Reentrant` flag then negate the default behavior?

I guess no, better to treat it as a bug, then figure out how to fix it in a non-breaking way.

Reply:
My thinking - if the exporter doesn't have the attribute at all, fall back to the old (and buggy) behavior, if the exporter has the attribute, start to enforce the correct behavior.